Add back history and reset subcommand in magics

akaihola opened the pull request:
- [x] documentation for `%config AiMagics.max_history`
- [x] manual testing for `%config AiMagics.max_history`
- [x] manual testing for limiting the maximum number of exchanges to include in the context
- [x] unit tests for all added features
- [ ] design decision about context for custom chains

#551 removed the history associated uniquely with the `openai-chat` provider in magic commands. It also removed the "reset" command to delete said history. Docs were updated to remove mention of the history and the reset command.

This PR adds back the history in magic commands. It also adds the `%ai reset` subcommand to delete said history. A mention of the history and the reset command are added to docs.

The history transcript maintains the distinction between human and AI messages by wrapping the prompts and responses in `HumanMessage` and `AIMessage` objects.

The maximum number of Human/AI message exchanges to include before the new prompt can be configured using `%config 
AiMagics.max_history = <n>`, defaulting to 2. The whole history is still kept in memory, so raising this option value will start including older messages in subsequent interactions.

For non-chat providers, human messages are ~~wrapped in pseudo XML `<HUMAN>...</HUMAN>` tags~~ prepended with `AI:` or `Human:` unless there is nothing but the first prompt. This is yet untested.

dlqqq commented on packages/jupyter-ai-magics/jupyter_ai_magics/magics.py:
```diff
@@ -437,6 +439,12 @@ def handle_error(self, args: ErrorArgs):
 
         return self.run_ai_cell(cell_args, prompt)
 
+    def _append_exchange(self, prompt: str, output: str):
+        """Appends a conversational exchange between user and an OpenAI Chat
+        model to a transcript that will be included in future exchanges."""
+        self.transcript.append(HumanMessage(prompt))
+        self.transcript.append(AIMessage(output))
+
```
There should be bounds on the length of the transcript passed to the LLM, since some LLMs have shorter token limits. Can you mimic the implementation of the history in chat, where we only allow up to 2 Human-AI exchanges? That would involve modifying this method to remove all but the last Human-AI exchange before appending the new Human-AI exchange to `self.transcript`.

If you wish to implement longer history as well, you can make the size of the history configurable via `%ai config`, while defaulting to 2 exchanges. See this PR for a reference: https://github.com/jupyterlab/jupyter-ai/pull/962

akaihola replied:
@dlqqq I assume you mean `%config AiMagics.<option> = <value>`, not `%ai config`? At least that's what [current documentation says](/jupyterlab/jupyter-ai/blob/main/docs/source/users/index.md#configuring-a-default-model).

I'll look into that!

dlqqq commented on packages/jupyter-ai-magics/jupyter_ai_magics/magics.py:
```diff
@@ -578,12 +589,29 @@ def run_ai_cell(self, args: CellArgs, prompt: str):
-            result = provider.generate([prompt])
+            if self.transcript:
+                transcript = [
+                    (
+                        f"<HUMAN>{message.content}</HUMAN>"
+                        if message.type == "human"
+                        else message.content
+                    )
+                    for message in self.transcript + [HumanMessage(content=prompt)]
```
I'm not familiar with this pseudo-XML syntax, and it may cause confuse LLMs to also return their responses in pseudo-XML, e.g. `<AI>The square root of pi is...</AI>`.

Can a more conventional string formatting be used instead here? LangChain simply prepends `Human: ` to human messages and `AI: ` to AI messages.

dlqqq reviewed (changes requested):
Thank you for opening this PR and for re-implementing message history in the magics! I've left suggestions about 1) bounding the chat history to 2 exchanges at most, and 2) avoiding the pseudo-XML syntax being used for non-chat providers. This is a good stopgap solution for users who want to use history in AI magics as soon as possible.

There are better ways to pass message history in LangChain however. In the future, we will definitely want to rework this logic to use the new LCEL syntax and use the `RunnableWithMessageHistory` class from `langchain_core.runnables.history`; see https://github.com/jupyterlab/jupyter-ai/issues/392.

dlqqq commented:
Thank you for your contribution! It's been a week since we've last heard from you, so I'm closing this PR due to inactivity. Please feel free to re-open this PR after addressing the review feedback.

akaihola commented:
> Thank you for your contribution! It's been a week since we've last heard from you, so I'm closing this PR due to inactivity. Please feel free to re-open this PR after addressing the review feedback.

I don't seem to have permission to re-open. Could you do that?

krassowski commented:
Yeah, that's the problem with closing PRs on GitHub.

akaihola commented:
Based on my manual testing, `max_history` works as expected when using `openrouter:openai/gpt-4o` as the model. If I prompt the model for my very first question, verbatim, it gives me the question `max_history` exchanges back.

Interestingly, if I use `openrouter:anthropic/claude-3.5-sonnet:beta` instead, the model remembers everything! So is there a history layer on OpenRouter side?!?

akaihola commented:
@krassowski & @dlqqq would it make sense to include message history in the context also when invoking custom chains? I haven't touched that part, so it still only sends the prompt entered in the cell:
```python
        if args.model_id in self.custom_model_registry and isinstance(
            self.custom_model_registry[args.model_id], LLMChain
        ):
            # Get the output, either as raw text or as the contents of the 'text' key of a dict
            invoke_output = self.custom_model_registry[args.model_id].invoke(prompt)
```

dlqqq commented on packages/jupyter-ai-magics/jupyter_ai_magics/magics.py:
```diff
@@ -577,13 +597,29 @@ def run_ai_cell(self, args: CellArgs, prompt: str):
         else:
             # generate output from model via provider
-            result = provider.generate([prompt])
+            if self.transcript:
```
The condition should be based on whether `context` exists, since `context` is used to build the input array, not `self.transcript`:

```suggestion
            if context:
```

dlqqq commented on packages/jupyter-ai-magics/jupyter_ai_magics/magics.py:
```diff
@@ -577,13 +597,29 @@ def run_ai_cell(self, args: CellArgs, prompt: str):
             # generate output from model via provider
-            result = provider.generate([prompt])
+            if self.transcript:
+                transcript = [
+                    (
+                        f"AI: {message.content}"
+                        if message.type == "ai"
+                        else f"{message.type.title()}: {message.content}"
+                    )
+                    for message in context + [HumanMessage(content=prompt)]
+                ]
+            else:
+                transcript = [prompt]
+            result = provider.generate(transcript)
```
Could you rename this local variable `transcript` to `inputs`? This should help avoid confusion with `self.transcript`.

```suggestion
                inputs = [
                    (
                        f"AI: {message.content}"
                        if message.type == "ai"
                        else f"{message.type.title()}: {message.content}"
                    )
                    for message in context + [HumanMessage(content=prompt)]
                ]
            else:
                inputs = [prompt]
            result = provider.generate(inputs)
```

dlqqq reviewed (changes requested):
@akaihola Awesome work! Thank you for including unit test coverage as well. I've left some additional minor feedback below. Once that's addressed, this is ready to be merged. 👍

akaihola commented:
@dlqqq I believe I've now addressed your fine review feedback!

dlqqq reviewed (approved):
@akaihola Awesome work, thank you! 🎉
